crates/nu-cli/src/types/deduction.rs

```diff
@@ -0,0 +1,1451 @@
+                    nu_protocol::hir::Literal::GlobPattern(_) => Some(SyntaxShape::String),
+                    nu_protocol::hir::Literal::Operator(_) => Some(SyntaxShape::Operator),
+                    nu_protocol::hir::Literal::ColumnPath(_) => Some(SyntaxShape::ColumnPath),
+                    nu_protocol::hir::Literal::Bare(_) => Some(SyntaxShape::String),
+                }
+            }
+            //Synthetic are expressions that are generated by the parser and not inputed by the user
+            //ExternalWord is anything sent to external commands (?)
+            Expression::ExternalWord => Some(SyntaxShape::String),
+            Expression::Synthetic(_) => Some(SyntaxShape::String),
+
+            Expression::Binary(_) => Some(SyntaxShape::Math),
+            Expression::Range(_) => Some(SyntaxShape::Range),
+            Expression::List(_) => Some(SyntaxShape::Table),
+            Expression::Boolean(_) => Some(SyntaxShape::String),
+
+            Expression::Path(_) => Some(SyntaxShape::ColumnPath),
+            Expression::FilePath(_) => Some(SyntaxShape::Path),
+            Expression::Block(_) => Some(SyntaxShape::Block),
+            Expression::ExternalCommand(_) => Some(SyntaxShape::String),
+            Expression::Table(_, _) => Some(SyntaxShape::Table),
+            Expression::Command => Some(SyntaxShape::String),
+            Expression::Invocation(_) => Some(SyntaxShape::Block),
+            Expression::Garbage => unreachable!("Should have failed at parsing stage"),
+        }
+    }
+
+    fn get_shape_of_binary_arg_or_insert_dependency(
+        &mut self,
+        //var depending on shape of expr (arg)
+        (var, expr): (&VarUsage, &SpannedExpression),
+        //source_bin is binary having var on one and expr on other side
+        source_bin: &SpannedExpression,
+        (pipeline_idx, pipeline): (usize, &Commands),
+        registry: &CommandRegistry,
+    ) -> Result<Option<SyntaxShape>, ShellError> {
+        trace!("Getting shape of binary arg {:?} for var {:?}", expr, var);
+        if let Some(shape) = self.get_shape_of_expr_or_insert_dependency(
+            var,
+            (op_of(source_bin), source_bin.span),
+            expr,
+        ) {
+            trace!("> Found shape: {:?}", shape);
+            Ok(match shape {
+                SyntaxShape::Math => {
+                    //If execution happens here, the situation is as follows:
+                    //There is an Binary expression (source_bin) with a variable on one side
+                    //and a binary (lets call it "deep binary") on the other:
+                    //e.G. $var + 1 * 1
+                    //Now we try to infer the shapes inside the deep binary, compute the resulting
+                    //shape based on the operator (see get_result_shape_of) and return that.
+                    //That won't work if one of the deeper binary left/right expr is a variable.
+                    //Then we insert an element into
```

Comment:
Can you say more about what it means that this inference won't work if there is a binary expression that has a variable deeper inside of it?